src/client/datascience/jupyter/kernels/kernel.ts

```diff
@@ -425,95 +377,71 @@ export class Kernel implements IKernel {
-                        this._onStatusChanged.fire('starting');
-                        this.notebook = await this.notebookProvider.createNotebook({
-                            document: this.notebookDocument,
-                            resource: this.resourceUri,
-                            ui: this.startupUI,
-                            kernelConnection: this.kernelConnectionMetadata,
-                            token: this.startCancellation.token
-                        });
-                        if (!this.notebook) {
-                            // This is an unlikely case.
-                            // getOrCreateNotebook would return undefined only if getOnly = true (an issue with typings).
-                            throw new Error('Kernel has not been started');
-                        }
-                        await this.initializeAfterStart(
-                            SysInfoReason.Start,
-                            this.notebookDocument,
-                            placeholderCellPromise
-                        );
-                    } catch (ex) {
-                        traceError(
-                            `failed to create INotebook in kernel, UI Disabled = ${this.startupUI.disableUI}`,
-                            ex
-                        );
-                        if (ex instanceof JupyterConnectError) {
-                            throw ex;
-                        }
-                        // Provide a user friendly message in case `ex` is some error thats not throw by us.
-                        const message = DataScience.sessionStartFailedWithKernel().format(
-                            getDisplayNameOrNameOfKernelConnection(this.kernelConnectionMetadata)
-                        );
-                        throw WrappedError.from(message + ' ' + ('message' in ex ? ex.message : ex.toString()), ex);
-                    }
-                    sendKernelTelemetryEvent(
-                        this.resourceUri,
-                        Telemetry.PerceivedJupyterStartupNotebook,
-                        stopWatch.elapsedTime
-                    );
-                    resolve(this.notebook);
-                    this._onStarted.fire();
-                    disposeAllDisposables(disposables);
-                } catch (ex) {
-                    disposeAllDisposables(disposables);
-                    sendKernelTelemetryEvent(
-                        this.resourceUri,
-                        Telemetry.NotebookStart,
-                        stopWatch.elapsedTime,
-                        undefined,
-                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                        ex as any
+            this._notebookPromise = this.createNotebook(new StopWatch())
+                .then((n) => {
+                    // If notebook session is disposed, make sure to clear the promise
+                    n.session.onDidDispose(
```

Comment:
This doesn't look right, session is a child component of Notebook.
If anything the Notebook should monitor the session and dispose itself when its child component session gets disposed.
This way, we don't need to monitor components n levels deep